File: delphi/headers/EasyPayBackend.pas
{ Payments and refunds to EasyPay backend }
unit EasyPayBackend;

interface

uses classes, System.SysUtils, EasyPayBackendImport, JetBeepTypes,
  VersionChecker;

type
  { Result of the payment operation }
  TEasyPayPaymentResult = record
    { Error string. If it's nil then the payment was successful }
    errorString: String;
    { Unique EasyPay transaction id }
    easyPayTransactionId: Integer;
    { Unique EasyPay payment request id }
    easyPayPaymentRequestUid: String;
  end;

  { Result of the refund operation }
  TEasyPayRefundResult = record
    { Error string. If it's nil then the refund was successful }
    errorString: String;
  end;

  { This handler is called once payment result is received }
  TEasyPayPaymentResultHandler = reference to procedure
    (result: TEasyPayPaymentResult);

  { This handler is called once refund result is received }
  TEasyPayRefundResultHandler = reference to procedure
    (result: TEasyPayRefundResult);

  { HTTPS calls to EasyPay API }
  TEasyPayBackend = class(TObject)
  private
    handle: TEasyPayBackendHandle;
  public
    { This handler is called once payment result is received }
    paymentResultHandler: TEasyPayPaymentResultHandler;

    { This handler is called once refund result is received }
    refundResultHandler: TEasyPayRefundResultHandler;

    { @param environment dev or production server environment
      @param merchantToken merchant's credentials
      @raises EJetBeepInvalidVersion Raised if jetbeep-*.dll version is not equal to *.pas version. IMPORTANT: Do not try to catch this exception as it means you're doing something wrong }
    constructor Create(environment: TEasyPayEnvironment; merchantKey: String);
    destructor Destroy; override;

    { Performs payment to EasyPay backend
      @param transactionId unique POS transaction id (should be the same value as provided to TAutoDevice.createPaymentToken)
      @param token payment token received from TAutoDevice
      @param amountInCoins amount in coins. E.g. for $1.05 should be 105. (should be the same value as provided to TAutoDevice.createPaymentToken)
      @param deviceId current deviceId. Could be obtained as TAutoDevice.deviceId
      @param cashierId unique POS id (should be the same value as provided to AutoDevice.createPaymentToken)
    }
    procedure MakePayment(merchantTransactionId: String; paymentToken: String;
      amountInCoins: Cardinal; deviceId: Cardinal; cashierId: String);

    { Performs Partials payment to EasyPay backend.
      @param transactionId unique POS transaction id (should be the same value as provided to AutoDevice.createPaymentToken)
      @param token payment token received from TAutoDevice
      @param amountInCoins amount in coins. E.g. for $1.05 should be 105. (should be the same value as provided to TAutoDevice.createPaymentToken)
      @param metadata additional information attached to the payment (payment partials recepients)
      @param deviceId current deviceId. Could be obtained as TAutoDevice.deviceId
      @param cashierId unique POS id (should be the same value as provided to TAutoDevice.createPaymentToken)
    }
    procedure MakePaymentPartials(merchantTransactionId: String;
      paymentToken: String; amountInCoins: Cardinal; deviceId: Cardinal;
      metadata: array of TMetadata; cashierId: String);

    { Performs refund operation to EasyPay backend
      @param easyPayTransactionId transaction id received in paymentResultHandler
      @param amountInCoins amount in coins. E.g. for $1.05 should be 105. (should be the same value as provided to TAutoDevice.createPaymentToken)
      @param deviceId current deviceId. Could be obtained as TAutoDevice.deviceId()
    }
    procedure MakeRefund(easyPayTransactionId: LongInt; amountInCoins: Cardinal;
      deviceId: Cardinal);

    { Perform refund operation for Partials payment to EasyPay backend</p>
      @param paymentRequestUid PaymentRequestUid received in paymentResultHandler
      @param amountInCoins amount in coins. E.g. for $1.05 should be 105. (should be the same value as provided to TAutoDevice.createPaymentToken)
      @param deviceId current deviceId. Could be obtained as TAutoDevice.deviceId() }
    procedure MakeRefundPartials(easyPayTransactionId: LongInt;
      amountInCoins: Cardinal; deviceId: Cardinal);
  end;

implementation

procedure CPaymentResultHandler(result: TCEasyPayPaymentResult; data: THandle);
  cdecl; forward;
procedure CRefundResultHandler(result: TCEasyPayRefundResult; data: THandle);
  cdecl; forward;

procedure CPaymentResultHandler(result: TCEasyPayPaymentResult; data: THandle);
var
  EasyPayBackend: TEasyPayBackend;
  ConvertedResult: TEasyPayPaymentResult;
begin
  EasyPayBackend := TEasyPayBackend(data);

  if EasyPayBackend.paymentResultHandler = nil then
  begin
    Exit;
  end;

  ConvertedResult.errorString := System.Utf8ToUnicodeString(result.errorString);
  ConvertedResult.easyPayTransactionId := result.easyPayTransactionId;
  ConvertedResult.easyPayPaymentRequestUid :=
    String(result.easyPayPaymentRequestUid);
  EasyPayBackend.paymentResultHandler(ConvertedResult);
end;

procedure CRefundResultHandler(result: TCEasyPayRefundResult; data: THandle);
var
  EasyPayBackend: TEasyPayBackend;
  ConvertedResult: TEasyPayRefundResult;
begin
  EasyPayBackend := TEasyPayBackend(data);

  if EasyPayBackend.paymentResultHandler = nil then
  begin
    Exit;
  end;

  ConvertedResult.errorString := System.Utf8ToUnicodeString(result.errorString);
  EasyPayBackend.refundResultHandler(ConvertedResult);
end;

constructor TEasyPayBackend.Create(environment: TEasyPayEnvironment;
  merchantKey: String);
begin
  TVersionChecker.check;
  inherited Create;
  handle := jetbeep_easypay_new(environment,
    PAnsiChar(AnsiString(merchantKey)));
end;

destructor TEasyPayBackend.Destroy;
begin
  jetbeep_easypay_free(handle);
  inherited;
end;

procedure TEasyPayBackend.MakePayment(merchantTransactionId: String;
  paymentToken: String; amountInCoins: Cardinal; deviceId: Cardinal;
  cashierId: String);
var
  error: TCJetBeepError;
begin
  error := jetbeep_easypay_make_payment(handle,
    PAnsiChar(AnsiString(merchantTransactionId)),
    PAnsiChar(AnsiString(paymentToken)), amountInCoins, deviceId,
    CPaymentResultHandler, THandle(Self), PAnsiChar(AnsiString(cashierId)));

  case error of
    JETBEEP_NO_ERROR:
      Exit;
  else
    raise EJetBeepIO.Create('System input-output exception');
  end;
end;

procedure TEasyPayBackend.MakePaymentPartials(merchantTransactionId: String;
  paymentToken: String; amountInCoins: Cardinal; deviceId: Cardinal;
  metadata: array of TMetadata; cashierId: String);
var
  error: TCJetBeepError;
begin
  error := jetbeep_easypay_make_payment_partials(handle,
    PAnsiChar(AnsiString(merchantTransactionId)),
    PAnsiChar(AnsiString(paymentToken)), amountInCoins, deviceId, @metadata[0],
    Length(metadata), CPaymentResultHandler, THandle(Self),
    PAnsiChar(AnsiString(cashierId)));

  case error of
    JETBEEP_NO_ERROR:
      Exit;
  else
    raise EJetBeepIO.Create('System input-output exception');
  end;
end;

procedure TEasyPayBackend.MakeRefund(easyPayTransactionId: LongInt;
  amountInCoins: Cardinal; deviceId: Cardinal);
var
  error: TCJetBeepError;
begin
  error := jetbeep_easypay_make_refund(handle, easyPayTransactionId,
    amountInCoins, deviceId, CRefundResultHandler, THandle(Self));

  case error of
    JETBEEP_NO_ERROR:
      Exit;
  else
    raise EJetBeepIO.Create('System input-output exception');
  end;
end;

procedure TEasyPayBackend.MakeRefundPartials(easyPayTransactionId: LongInt;
  amountInCoins: Cardinal; deviceId: Cardinal);
var
  error: TCJetBeepError;
begin
  error := jetbeep_easypay_make_refund_partials(handle, easyPayTransactionId,
    amountInCoins, deviceId, CRefundResultHandler, THandle(Self));

  case error of
    JETBEEP_NO_ERROR:
      Exit;
  else
    raise EJetBeepIO.Create('System input-output exception');
  end;
end;

end.

File: lib/c-binding/easypay_backend.cpp
#define JETBEEP_API_EXPORTS
#include "../libjetbeep.h"
#include "../libjetbeep.hpp"

using namespace JetBeep;
using namespace std;

JETBEEP_API jetbeep_easypay_handle_t jetbeep_easypay_new(jetbeep_easypay_environment_t environment, const char* merchant_secret_key) {
  auto secretKey = string(merchant_secret_key);
  return new EasyPayBackend((EasyPayHostEnv)environment, secretKey);
}

JETBEEP_API void jetbeep_easypay_free(jetbeep_easypay_handle_t handle) {
  auto backend = (EasyPayBackend*)handle;
  delete backend;
}

JETBEEP_API jetbeep_error_t jetbeep_easypay_make_payment(jetbeep_easypay_handle_t handle,
                                                         const char* merchant_transaction_id,
                                                         const char* payment_token,
                                                         uint32_t amount_in_coins,
                                                         uint32_t device_id,
                                                         jetbeep_easypay_payment_result_cb callback,
                                                         void* data,
                                                         const char* cashier_id) {
  auto backend = (EasyPayBackend*)handle;
  try {
    auto merchantTransactionId = string(merchant_transaction_id);
    auto paymentToken = string(payment_token);
    auto cashierId = string(cashier_id);

    backend->makePayment(merchantTransactionId, paymentToken, amount_in_coins, device_id, cashierId)
      .then([callback, data](EasyPayResult result) {
        jetbeep_easypay_payment_result_t payment_result;
        if (result.isError()) {
          memset(&payment_result, 0, sizeof(payment_result));
          payment_result.error_string = result.primaryErrorMsg.c_str();
        } else {
          payment_result.easypay_payment_request_uid = result.PaymentRequestUid.c_str();
          payment_result.easypay_transaction_id = result.TransactionId;
          payment_result.error_string = nullptr;
        }
        callback(payment_result, data);
      })
      .catchError([callback, data](const exception_ptr&) {
        jetbeep_easypay_payment_result_t payment_result;
        memset(&payment_result, 0, sizeof(payment_result));
        payment_result.error_string = "Network error";
        callback(payment_result, data);
      });
  } catch (...) {
    return JETBEEP_ERROR_IO;
  }
  return JETBEEP_NO_ERROR;
}

JETBEEP_API jetbeep_error_t jetbeep_easypay_make_payment_partials(jetbeep_easypay_handle_t handle,
                                                                  const char* merchant_transaction_id,
                                                                  const char* payment_token,
                                                                  uint32_t amount_in_coins,
                                                                  uint32_t device_id,
                                                                  jetbeep_payment_metadata_t* metadata,
                                                                  size_t metadata_size,
                                                                  jetbeep_easypay_payment_result_cb callback,
                                                                  void* data,
                                                                  const char* cashier_id) {
  auto backend = (EasyPayBackend*)handle;
  try {
    auto merchantTransactionId = string(merchant_transaction_id);
    auto paymentToken = string(payment_token);
    auto cashierId = string(cashier_id);
    PaymentMetadata metaData;

    for (size_t i = 0; i < metadata_size; ++i) {
      auto key = string(metadata[i].key);
      auto value = string(metadata[i].value);
      metaData[key] = value;
    }

    backend->makePaymentPartials(merchantTransactionId, paymentToken, amount_in_coins, device_id, metaData, cashierId)
      .then([callback, data](EasyPayResult result) {
        jetbeep_easypay_payment_result_t payment_result;
        if (result.isError()) {
          memset(&payment_result, 0, sizeof(payment_result));
          payment_result.error_string = result.primaryErrorMsg.c_str();
        } else {
          payment_result.easypay_payment_request_uid = result.PaymentRequestUid.c_str();
          payment_result.easypay_transaction_id = result.TransactionId;
          payment_result.error_string = nullptr;
        }
        callback(payment_result, data);
      })
      .catchError([callback, data](const exception_ptr&) {
        jetbeep_easypay_payment_result_t payment_result;
        memset(&payment_result, 0, sizeof(payment_result));
        payment_result.error_string = "Network error";
        callback(payment_result, data);
      });
  } catch (...) {
    return JETBEEP_ERROR_IO;
  }
  return JETBEEP_NO_ERROR;
}

JETBEEP_API jetbeep_error_t jetbeep_easypay_make_refund(jetbeep_easypay_handle_t handle,
                                                        long easypay_transaction_id,
                                                        uint32_t amount_in_coins,
                                                        uint32_t device_id,
                                                        jetbeep_easypay_refund_result_cb callback,
                                                        void* data) {
  auto backend = (EasyPayBackend*)handle;
  try {
    backend->makeRefund(easypay_transaction_id, amount_in_coins, device_id)
      .then([callback, data](EasyPayResult result) {
        jetbeep_easypay_refund_result_t refund_result;

        if (result.isError()) {
          refund_result.error_string = result.primaryErrorMsg.c_str();
        } else {
          refund_result.error_string = nullptr;
        }
        callback(refund_result, data);
      })
      .catchError([callback, data](exception_ptr) {
        jetbeep_easypay_refund_result_t refund_result;

        refund_result.error_string = "Network error";
        callback(refund_result, data);
      });
  } catch (...) {
    return JETBEEP_ERROR_IO;
  }
  return JETBEEP_NO_ERROR;
}

JETBEEP_API jetbeep_error_t jetbeep_easypay_make_refund_partials(jetbeep_easypay_handle_t handle,
                                                                 const char* payment_request_uid,
                                                                 uint32_t amount_in_coins,
                                                                 uint32_t device_id,
                                                                 jetbeep_easypay_refund_result_cb callback,
                                                                 void* data) {
  auto backend = (EasyPayBackend*)handle;
  try {
    auto paymentRequestId = string(payment_request_uid);
    backend->makeRefundPartials(payment_request_uid, amount_in_coins, device_id)
      .then([callback, data](EasyPayResult result) {
        jetbeep_easypay_refund_result_t refund_result;

        if (result.isError()) {
          refund_result.error_string = result.primaryErrorMsg.c_str();
        } else {
          refund_result.error_string = nullptr;
        }
        callback(refund_result, data);
      })
      .catchError([callback, data](exception_ptr) {
        jetbeep_easypay_refund_result_t refund_result;

        refund_result.error_string = "Network error";
        callback(refund_result, data);
      });
  } catch (...) {
    return JETBEEP_ERROR_IO;
  }
  return JETBEEP_NO_ERROR;
}